File: implementations/go/binary_search/small/binary_search_small.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const N = 100

var arr [N]int64

func setup() {
	for i := 0; i < N; i++ {
		arr[i] = int64(i) * 2
	}
}

func binarySearch() int {
	target := int64(N-1) * 2
	lo, hi := 0, N-1
	for lo <= hi {
		mid := lo + (hi-lo)/2
		if arr[mid] == target {
			return mid
		} else if arr[mid] < target {
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}
	return -1
}

func main() {
	setup()

	fmt.Println("ready")

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		fmt.Println(binarySearch())
	}
}

File: implementations/rust/binary_search/small/binary_search_small.rs
use std::io::{self, BufRead, Write};

const N: usize = 100;

static mut ARR: [i64; N] = [0; N];

fn setup() {
    unsafe {
        for i in 0..N {
            ARR[i] = (i as i64) * 2;
        }
    }
}

fn binary_search() -> i64 {
    unsafe {
        let target = (N as i64 - 1) * 2;
        let mut lo = 0i64;
        let mut hi = N as i64 - 1;
        while lo <= hi {
            let mid = lo + (hi - lo) / 2;
            if ARR[mid as usize] == target {
                return mid;
            } else if ARR[mid as usize] < target {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        -1
    }
}

fn main() {
    setup();

    println!("ready");
    io::stdout().flush().unwrap();

    let stdin = io::stdin();
    for _ in stdin.lock().lines() {
        println!("{}", binary_search());
        io::stdout().flush().unwrap();
    }
}